packages/core/src/shared/utilities/map.ts

```diff
@@ -0,0 +1,98 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getLogger } from '../logger'
+
+/**
+ * Simply maps a Key to an Object. Everything is done in memory, allowing for atomic updates.
+ *
+ * There is a common use case for this class, so look to reuse this when possible.
+ * Benefits, instead of making your own:
+ * - A key does not have to be a string, {@link MemoryMap.asKey} enables this
+ * - Implements most of the boilerplate
+ */
+export abstract class MemoryMap<Key, Value = { [key: string]: any }> implements MapSync<Key, Value> {
+    private readonly data: { [key: string]: Value | undefined } = {}
+
+    exists(key: Key): boolean {
+        return this.data[this.asKey(key)] !== undefined
+    }
+
+    /**
+     * IMPORTANT: If getting a key without having {@link set()} a value, the {@link default} will be returned.
+     * Use {@link exists()} to check for existence before calling get(), if necessary.
+     */
+    get(key: Key): Value {
+        const actualKey = this.asKey(key)
+        const result = this.data[actualKey]
+        return result ?? this.default
+    }
+
+    set(key: Key, data: Partial<Value>): void {
+        const currentData = this.get(key)
+        this.data[this.asKey(key)] = { ...currentData, ...data }
+    }
+
+    clear(key: Key, reason: string): void {
+        delete this.data[this.asKey(key)]
+        getLogger().debug(`${this.name}: cleared cache, key: ${JSON.stringify(key)}, reason: ${reason}`)
+    }
+
+    /**
+     * Converts the user-provided key to a string so that it can be used
+     * as an object key.
+     */
+    protected abstract asKey(key: Key): string
```

Comment:
seems to overlap with `TelemetryDebounceInfo` https://github.com/aws/aws-toolkit-vscode/blob/d49eee404d1d65332c955c55d907116ae870e378/packages/core/src/shared/vscode/commands2.ts#L605

can we reconcile them? maybe we should start with the interface defined by `TelemetryDebounceInfo` and see what the delta is.

Reply:
This makes sense. I don't want it to block this PR anymore. I'll look to do this after we merge.